### PR TITLE
Added minimum time threshold when triggering connection lost

### PR DESCRIPTION
**Pull request by eabbott**
Currently the spymemcached drivers will detect a lost memcached server in two ways. Both methods appear too slow in practice.

After 996 consecutive timed out memcached calls will trigger a server as lost. The timeout is 1s, so a delay is built in depending on the concurrency. This method of detection will take more or less time depending on the particular traffic profile of the application. Unfortunately even our most highly trafficked service capiweb takes several minutes [to detect failure.](https://hubspot.tv.appneta.com/app/ContactsApiWeb/app_server/visualize#fs=1402992720&fe=1402996320&v=heatmap&xpd=1353390&ts=1402994400000&te=1402994640000&nd=912508&xd=1151943)

The second method of detection is a bit fuzzy and may only exist as an artifact of testing. During functional testing on my laptop/aws with iptables used to simulate a failure, the lost server is detected in 20-40s. This method of detection has not been seen in and environment with real traffic such as qa or production. Even in qa when iptables has been used to simulate failure, the failed server is not detected for several minutes. This implies that load is a factor. Again I will stress the unknown and that we really haven't done many tests yet in qa/prod, so the data is very sparse.

This PR introduces detection determinism based upon the clock. With this PR we can configure a memcached pool to detect a host as failed after X consecutive timed out requests (eg 50) over N millis (eg 30sec). The goal is to detect a memcached failure in a deterministic time period (eg 30s).

As for the code changes themselves, the particular angle taken is one of eventual consistency. The sync aspects are not at all exact, there is leeway for things to be missed immediately, but eventually they will be picked up correctly. To say it more clearly in prose, it's entirely possible that exactness will be missed and detection will happen ms to sec later than it could have. That is deemed an acceptable improvement over the current state of the drivers. In particular I'm referring to TCPMemcachedNodeImpl.continuousTimeout and TCPMemcachedNodeImpl.continuousTimeoutStart being synchronized separately.

Our version of the drivers are 3yrs out of date, the up to date version does not have similar functionality. Its feasible that if this code proves useful that we could submit a PR.

@HiJon89 @axiak @marlier @mjball @wsorenson @jaredstehler

### CODE REVIEW

File: src/main/java/net/spy/memcached/MemcachedConnection.java [hunk 1]
```diff
@@ -282,9 +284,10 @@ public void handleIO() throws IOException {
 		// see if any connections blew up with large number of timeouts
 		for(SelectionKey sk : selector.keys()) {
 			MemcachedNode mn = (MemcachedNode)sk.attachment();
-			if (mn.getContinuousTimeout() > timeoutExceptionThreshold)
+			if (mn.getContinuousTimeout() > timeoutExceptionThreshold &&
+					(System.nanoTime() - mn.getContinuousTimeoutStart())/1000000 > timeoutExceptionDurationThreshold)
```
Comment (sit) [inline src/main/java/net/spy/memcached/MemcachedConnection.java]: I might extract a function for this `if` condition.

File: src/main/java/net/spy/memcached/protocol/TCPMemcachedNodeImpl.java [hunk 2]
```diff
@@ -502,9 +506,13 @@ public final int writeSome() throws IOException {
...
 	public void setContinuousTimeout(boolean timedOut) {
 		if (timedOut && isActive()) {
-			continuousTimeout.incrementAndGet();
+			int count = continuousTimeout.incrementAndGet();
+			if (count == 1) {
+				continuousTimeoutStart.set(System.nanoTime());
+			}
 		} else {
 			continuousTimeout.set(0);
+			continuousTimeoutStart.set(0);
```
Comment (sit) [inline src/main/java/net/spy/memcached/protocol/TCPMemcachedNodeImpl.java]: Could we have ...

thread A: continuousTimeout.set(0) (line 514)
thread B: continuousTimeout.incrementAndGet() (line 509)
thread B: continuousTimeoutStart.set(System.nanoTime()); (line 510)
thread A: continuousTimeoutStart.set(0) (line 515)

Seems like we ought to actually extract a class to track timeouts that combines the timeout and timeoutstart and protects the whole thing with a lock.

File: src/main/java/net/spy/memcached/MemcachedConnection.java (hunk 1 above)
Comment (sit) [inline src/main/java/net/spy/memcached/MemcachedConnection.java]: Here I think we could have a race condition where thread A is in here and thread B increments continuousTimeout but has not set continuousTimeoutStart (which would thus be 0) so that thread A sees `System.nanoTime() - 0) > timeoutExceptionDurationThreshold.`

Comment (eabbott) [inline src/main/java/net/spy/memcached/MemcachedConnection.java]: The if statement is definitely a bit wordy, but it fits with the rest of the code style.

**Comment by eabbott** [comment]
For your comments about the sync @sit , I agree and have attempted to address the issue in the PR part about the code being "eventually consistent". Did you read that part and simply disagree with my approach? I argued that the desired behavior will work fine without the "correct" synchronization.

**Comment by eabbott** [comment]
Thanks @sit! This version condenses the two locks into one to avoid the initial race condition.

**Comment by sit** [comment]
Cool.

I might want to scope down the lock more (e.g., a specific lock object) because I don't know if anyone elsewhere might be synchronizing on the node object.

That aside, :+1:

**Comment by sit** [comment]
:shipit:
